Manifold/Constraint.swift

```diff
@@ -32,3 +59,78 @@ public func == (left: Constraint, right: Constraint) -> Bool {
 public func === (left: Type, right: Type) -> Constraint {
 	return Constraint(equality: left, right)
 }
+
+
+// MARK: - ConstraintSet
+
+public typealias ConstraintSet = Multiset<Constraint>
+
+private func typeGraph(constraints: ConstraintSet) -> (DisjointSet<Type>, [Type: Int]) {
+	let distinctTypes = Set(lazy(constraints)
+		.flatMap {
+			$0.analysis(ifEquality: {
+				$0.instantiate().distinctTypes.union($1.instantiate().distinctTypes)
+			})
+		})
+	let typeGraph = DisjointSet(distinctTypes)
+	return (typeGraph, Dictionary(lazy(enumerate(typeGraph)).map { ($1, $0) }))
+}
```

Comment:
This constructs a `DisjointSet` and a `Dictionary` of types to the `DisjointSet`’s indices.